File: src/primitives/address.test.ts
import { describe, test, expect } from "bun:test";
import {
  Address,
  ZERO_ADDRESS,
  validateChecksum,
  calculateCreateAddress,
  calculateCreate2Address,
} from "./address.ts";

describe("Address creation", () => {
  test("fromHex - valid address", () => {
    const addr = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    expect(addr.bytes.length).toBe(20);
  });

  test("fromHex - with 0x prefix", () => {
    const addr = Address.fromHex(
      "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"
    );
    expect(addr.toHex()).toBe("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266");
  });

  test("fromHex - mixed case", () => {
    const addr = Address.fromHex(
      "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"
    );
    expect(addr.toHex()).toBe("0xa0cf798816d4b9b9866b5330eea46a18382f251e");
  });

  test("fromHex - rejects invalid length", () => {
    expect(() => Address.fromHex("0x123")).toThrow();
    expect(() => Address.fromHex("0x" + "00".repeat(21))).toThrow();
  });

  test("fromHex - rejects invalid characters", () => {
    expect(() => Address.fromHex("0x" + "zz".repeat(20))).toThrow();
  });

  test("fromHex - rejects missing 0x", () => {
    expect(() => Address.fromHex("a0cf798816d4b9b9866b5330eea46a18382f251e")).toThrow();
  });
});

describe("Address conversion", () => {
  test("toHex - lowercase output", () => {
    const addr = Address.fromHex(
      "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"
    );
    expect(addr.toHex()).toBe("0xa0cf798816d4b9b9866b5330eea46a18382f251e");
  });

  test("toChecksumHex - EIP-55 checksum", () => {
    const testCases = [
      {
        input: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
        expected: "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
      },
      {
        input: "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
        expected: "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
      },
      {
        input: "0x70997970c51812dc3a010c7d01b50e0d17dc79c8",
        expected: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
      },
    ];

    for (const { input, expected } of testCases) {
      const addr = Address.fromHex(input);
      expect(addr.toChecksumHex()).toBe(expected);
    }
  });
});

describe("Address validation", () => {
  test("validateChecksum - valid checksums", () => {
    expect(
      validateChecksum("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e")
    ).toBe(true);
    expect(
      validateChecksum("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266")
    ).toBe(true);
    expect(
      validateChecksum("0x70997970C51812dc3A010C7d01b50e0d17dc79C8")
    ).toBe(true);
  });

  test("validateChecksum - invalid checksums", () => {
    expect(
      validateChecksum("0xa0cf798816d4b9b9866b5330eea46a18382f251e")
    ).toBe(false);
    expect(
      validateChecksum("0xA0CF798816D4B9B9866B5330EEA46A18382F251E")
    ).toBe(false);
  });
});

describe("Address equality", () => {
  test("equals - same address", () => {
    const addr1 = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const addr2 = Address.fromHex(
      "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"
    );
    expect(addr1.equals(addr2)).toBe(true);
  });

  test("equals - different address", () => {
    const addr1 = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const addr2 = Address.fromHex(
      "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"
    );
    expect(addr1.equals(addr2)).toBe(false);
  });

  test("isZero - zero address", () => {
    const zero = Address.fromHex("0x" + "00".repeat(20));
    expect(zero.isZero()).toBe(true);
    expect(ZERO_ADDRESS.isZero()).toBe(true);
  });

  test("isZero - non-zero address", () => {
    const addr = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    expect(addr.isZero()).toBe(false);
  });
});

describe("Address from public key", () => {
  test("fromPublicKey - derives correct address", () => {
    // Test vector from Ethereum: Anvil account #0
    const pubKeyX =
      0x8318535b54105d4a7aae60c08fc45f9687181b4fdfc625bd1a753fa7397fed75n;
    const pubKeyY =
      0x3547f11ca8696646f2f3acb08e31016afac23e630c5d11f59f61fef57b0d2aa5n;

    const addr = Address.fromPublicKey(pubKeyX, pubKeyY);
    expect(addr.toChecksumHex()).toBe(
      "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
    );
  });
});

describe("Address numeric conversion", () => {
  test("toU256 and fromU256 - round trip", () => {
    const addr = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const u256 = addr.toU256();
    const result = Address.fromU256(u256);
    expect(result.equals(addr)).toBe(true);
  });

  test("fromU256 - uses last 20 bytes", () => {
    const value = 0xffffffffffffffffffffffffffffffffffffffffn;
    const addr = Address.fromU256(value);
    expect(addr.toHex()).toBe("0x" + "ff".repeat(20));
  });
});

describe("CREATE address calculation", () => {
  test("calculateCreateAddress - nonce 0", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const addr = calculateCreateAddress(deployer, 0n);
    expect(addr.toHex()).toBe("0xcd234a471b72ba2f1ccf0a70fcaba648a5eecd8d");
  });

  test("calculateCreateAddress - nonce 1", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const addr = calculateCreateAddress(deployer, 1n);
    expect(addr.toHex()).toBe("0x343c43a37d37dff08ae8c4a11544c718abb4fcf8");
  });

  test("calculateCreateAddress - various nonces", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );

    const testCases = [
      { nonce: 0n, expected: "0xcd234a471b72ba2f1ccf0a70fcaba648a5eecd8d" },
      { nonce: 1n, expected: "0x343c43a37d37dff08ae8c4a11544c718abb4fcf8" },
      { nonce: 2n, expected: "0xf778b86fa74e846c4f0a1fbd1335fe81c00a0c91" },
      { nonce: 3n, expected: "0xfffd933a0bc612844eaf0c6fe3e5b8e9b6c1d19c" },
    ];

    for (const { nonce, expected } of testCases) {
      const addr = calculateCreateAddress(deployer, nonce);
      expect(addr.toHex()).toBe(expected);
    }
  });

  test("calculateCreateAddress - deterministic", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const addr1 = calculateCreateAddress(deployer, 42n);
    const addr2 = calculateCreateAddress(deployer, 42n);
    expect(addr1.equals(addr2)).toBe(true);
  });
});

describe("CREATE2 address calculation", () => {
  test("calculateCreate2Address - zero values", () => {
    const deployer = Address.fromHex("0x" + "00".repeat(20));
    const salt = new Uint8Array(32);
    const initCodeHash = new Uint8Array(32);

    const addr = calculateCreate2Address(deployer, salt, initCodeHash);
    expect(addr.toHex()).toBe("0x4d1a2e2bb4f88f0250f26ffff098b0b30b26bf38");
  });

  test("calculateCreate2Address - deterministic", () => {
    const deployer = Address.fromHex(
      "0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef"
    );
    const salt = new Uint8Array(32);
    salt[0] = 0x12;
    salt[1] = 0x34;
    salt[2] = 0x56;
    salt[3] = 0x78;

    const initCodeHash = new Uint8Array(32);
    initCodeHash[0] = 0xab;
    initCodeHash[1] = 0xcd;
    initCodeHash[2] = 0xef;

    const addr1 = calculateCreate2Address(deployer, salt, initCodeHash);
    const addr2 = calculateCreate2Address(deployer, salt, initCodeHash);
    expect(addr1.equals(addr2)).toBe(true);
  });

  test("calculateCreate2Address - different with different deployer", () => {
    const deployer1 = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const deployer2 = Address.fromHex(
      "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"
    );
    const salt = new Uint8Array(32);
    const initCodeHash = new Uint8Array(32);

    const addr1 = calculateCreate2Address(deployer1, salt, initCodeHash);
    const addr2 = calculateCreate2Address(deployer2, salt, initCodeHash);
    expect(addr1.equals(addr2)).toBe(false);
  });

  test("calculateCreate2Address - different with different salt", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const salt1 = new Uint8Array(32);
    salt1[0] = 0x01;
    const salt2 = new Uint8Array(32);
    salt2[0] = 0x02;
    const initCodeHash = new Uint8Array(32);

    const addr1 = calculateCreate2Address(deployer, salt1, initCodeHash);
    const addr2 = calculateCreate2Address(deployer, salt2, initCodeHash);
    expect(addr1.equals(addr2)).toBe(false);
  });

  test("calculateCreate2Address - different with different initCodeHash", () => {
    const deployer = Address.fromHex(
      "0xa0cf798816d4b9b9866b5330eea46a18382f251e"
    );
    const salt = new Uint8Array(32);
    const initCodeHash1 = new Uint8Array(32);
    initCodeHash1[0] = 0x01;
    const initCodeHash2 = new Uint8Array(32);
    initCodeHash2[0] = 0x02;

    const addr1 = calculateCreate2Address(deployer, salt, initCodeHash1);
    const addr2 = calculateCreate2Address(deployer, salt, initCodeHash2);
    expect(addr1.equals(addr2)).toBe(false);
  });
});

describe("Edge cases", () => {
  test("Address - all zeros", () => {
    const zero = Address.fromHex("0x" + "00".repeat(20));
    expect(zero.isZero()).toBe(true);
    expect(zero.toHex()).toBe("0x" + "00".repeat(20));
  });

  test("Address - all ones", () => {
    const ones = Address.fromHex("0x" + "ff".repeat(20));
    expect(ones.isZero()).toBe(false);
    expect(ones.toHex()).toBe("0x" + "ff".repeat(20));
  });
});

File: primitives/address_test.go
package primitives

import (
	"math/big"
	"testing"
)

func TestAddressFromHex(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		wantErr bool
	}{
		{
			name:    "valid lowercase",
			input:   "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			wantErr: false,
		},
		{
			name:    "valid checksum",
			input:   "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
			wantErr: false,
		},
		{
			name:    "invalid length",
			input:   "0xa0cf79",
			wantErr: true,
		},
		{
			name:    "missing prefix",
			input:   "a0cf798816d4b9b9866b5330eea46a18382f251e",
			wantErr: true,
		},
		{
			name:    "invalid character",
			input:   "0xa0cf798816d4b9b9866b5330eea46a18382f251g",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := AddressFromHex(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("AddressFromHex(%q) error = %v, wantErr %v", tt.input, err, tt.wantErr)
				return
			}
			if !tt.wantErr && addr.IsZero() && tt.input != "0x0000000000000000000000000000000000000000" {
				t.Errorf("AddressFromHex(%q) returned zero address unexpectedly", tt.input)
			}
		})
	}
}

func TestAddressToHex(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "lowercase address",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			expected: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
		},
		{
			name:     "zero address",
			input:    "0x0000000000000000000000000000000000000000",
			expected: "0x0000000000000000000000000000000000000000",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := AddressFromHex(tt.input)
			if err != nil {
				t.Fatalf("AddressFromHex(%q) failed: %v", tt.input, err)
			}
			result := addr.ToHex()
			if result != tt.expected {
				t.Errorf("ToHex() = %s, want %s", result, tt.expected)
			}
		})
	}
}

func TestAddressToChecksumHex(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "test case 1",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			expected: "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
		},
		{
			name:     "test case 2",
			input:    "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
			expected: "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
		},
		{
			name:     "test case 3",
			input:    "0x70997970c51812dc3a010c7d01b50e0d17dc79c8",
			expected: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := AddressFromHex(tt.input)
			if err != nil {
				t.Fatalf("AddressFromHex(%q) failed: %v", tt.input, err)
			}
			result := addr.ToChecksumHex()
			if result != tt.expected {
				t.Errorf("ToChecksumHex() = %s, want %s", result, tt.expected)
			}
		})
	}
}

func TestValidateChecksum(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected bool
	}{
		{
			name:     "valid checksum",
			input:    "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
			expected: true,
		},
		{
			name:     "invalid checksum (all lowercase)",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			expected: false,
		},
		{
			name:     "invalid checksum (all uppercase)",
			input:    "0xA0CF798816D4B9B9866B5330EEA46A18382F251E",
			expected: false,
		},
		{
			name:     "valid checksum 2",
			input:    "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
			expected: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ValidateChecksum(tt.input)
			if result != tt.expected {
				t.Errorf("ValidateChecksum(%q) = %v, want %v", tt.input, result, tt.expected)
			}
		})
	}
}

func TestAddressFromPublicKey(t *testing.T) {
	// Test vectors from Zig tests
	publicKeyX := new(big.Int)
	publicKeyX.SetString("8318535b54105d4a7aae60c08fc45f9687181b4fdfc625bd1a753fa7397fed75", 16)

	publicKeyY := new(big.Int)
	publicKeyY.SetString("3547f11ca8696646f2f3acb08e31016afac23e630c5d11f59f61fef57b0d2aa5", 16)

	addr := AddressFromPublicKey(publicKeyX, publicKeyY)

	expected, err := AddressFromHex("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266")
	if err != nil {
		t.Fatalf("Failed to create expected address: %v", err)
	}

	if !addr.Equals(expected) {
		t.Errorf("AddressFromPublicKey() = %s, want %s", addr.ToHex(), expected.ToHex())
	}

	// Verify checksum
	checksummed := addr.ToChecksumHex()
	expectedChecksum := "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
	if checksummed != expectedChecksum {
		t.Errorf("ToChecksumHex() = %s, want %s", checksummed, expectedChecksum)
	}
}

func TestAddressIsZero(t *testing.T) {
	zeroAddr := Address{}
	if !zeroAddr.IsZero() {
		t.Error("Zero address should return true for IsZero()")
	}

	nonZeroAddr, _ := AddressFromHex("0xa0cf798816d4b9b9866b5330eea46a18382f251e")
	if nonZeroAddr.IsZero() {
		t.Error("Non-zero address should return false for IsZero()")
	}
}

func TestAddressEquals(t *testing.T) {
	addr1, _ := AddressFromHex("0xa0cf798816d4b9b9866b5330eea46a18382f251e")
	addr2, _ := AddressFromHex("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e")
	addr3, _ := AddressFromHex("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266")

	if !addr1.Equals(addr2) {
		t.Error("Same addresses should be equal regardless of case")
	}

	if addr1.Equals(addr3) {
		t.Error("Different addresses should not be equal")
	}
}

func TestIsValidAddress(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected bool
	}{
		{
			name:     "valid lowercase",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			expected: true,
		},
		{
			name:     "valid checksum",
			input:    "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
			expected: true,
		},
		{
			name:     "too short",
			input:    "0xa0cf79",
			expected: false,
		},
		{
			name:     "no prefix",
			input:    "a0cf798816d4b9b9866b5330eea46a18382f251e",
			expected: false,
		},
		{
			name:     "invalid character",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251z",
			expected: false,
		},
		{
			name:     "too long",
			input:    "0xa0cf798816d4b9b9866b5330eea46a18382f251eff",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := IsValidAddress(tt.input)
			if result != tt.expected {
				t.Errorf("IsValidAddress(%q) = %v, want %v", tt.input, result, tt.expected)
			}
		})
	}
}

func TestCalculateCreateAddress(t *testing.T) {
	tests := []struct {
		name     string
		deployer string
		nonce    uint64
		expected string
	}{
		{
			name:     "nonce 0",
			deployer: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			nonce:    0,
			expected: "0xcd234a471b72ba2f1ccf0a70fcaba648a5eecd8d",
		},
		{
			name:     "nonce 1",
			deployer: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			nonce:    1,
			expected: "0x343c43a37d37dff08ae8c4a11544c718abb4fcf8",
		},
		{
			name:     "nonce 2",
			deployer: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			nonce:    2,
			expected: "0xf778b86fa74e846c4f0a1fbd1335fe81c00a0c91",
		},
		{
			name:     "nonce 3",
			deployer: "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
			nonce:    3,
			expected: "0xfffd933a0bc612844eaf0c6fe3e5b8e9b6c1d19c",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			deployer, err := AddressFromHex(tt.deployer)
			if err != nil {
				t.Fatalf("Failed to parse deployer address: %v", err)
			}

			result, err := CalculateCreateAddress(deployer, tt.nonce)
			if err != nil {
				t.Fatalf("CalculateCreateAddress() error = %v", err)
			}

			expected, err := AddressFromHex(tt.expected)
			if err != nil {
				t.Fatalf("Failed to parse expected address: %v", err)
			}

			if !result.Equals(expected) {
				t.Errorf("CalculateCreateAddress() = %s, want %s", result.ToHex(), expected.ToHex())
			}
		})
	}
}

func TestCalculateCreate2Address(t *testing.T) {
	// Test with zero values
	deployer := Address{}
	salt := [32]byte{}
	initCodeHash := [32]byte{}

	result := CalculateCreate2Address(deployer, salt, initCodeHash)
	expected, _ := AddressFromHex("0x4d1a2e2bb4f88f0250f26ffff098b0b30b26bf38")

	if !result.Equals(expected) {
		t.Errorf("CalculateCreate2Address() = %s, want %s", result.ToHex(), expected.ToHex())
	}

	// Test deterministic: same inputs should give same output
	result2 := CalculateCreate2Address(deployer, salt, initCodeHash)
	if !result.Equals(result2) {
		t.Error("CalculateCreate2Address() should be deterministic")
	}
}

func TestAddressFromU256AndToU256(t *testing.T) {
	value := big.NewInt(0xdeadbeef)
	addr := AddressFromU256(value)

	// Convert back to U256
	result := addr.ToU256()

	if result.Cmp(value) != 0 {
		t.Errorf("Round trip failed: got %s, want %s", result.String(), value.String())
	}
}

func TestAddressBytes(t *testing.T) {
	addr, _ := AddressFromHex("0xa0cf798816d4b9b9866b5330eea46a18382f251e")
	bytes := addr.Bytes()

	if len(bytes) != 20 {
		t.Errorf("Bytes() length = %d, want 20", len(bytes))
	}

	// Verify first few bytes
	expected := []byte{0xa0, 0xcf, 0x79, 0x88}
	for i := 0; i < 4; i++ {
		if bytes[i] != expected[i] {
			t.Errorf("Bytes()[%d] = 0x%02x, want 0x%02x", i, bytes[i], expected[i])
		}
	}
}

func TestAddressString(t *testing.T) {
	addr, _ := AddressFromHex("0xa0cf798816d4b9b9866b5330eea46a18382f251e")
	str := addr.String()

	// String() should return checksummed address
	expected := "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"
	if str != expected {
		t.Errorf("String() = %s, want %s", str, expected)
	}
}
